File: java/src/main/java/com/quant/frtb/Validation.java
package com.quant.frtb;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;

/**
 * Independent model validation framework.
 *
 * <p>Checks (all pinned):
 * <ul>
 *   <li>Benchmarking — project BS pricer vs an independent CRR binomial
 *       lattice (501 steps) on a pinned strike/maturity/call-put grid;
 *       PASS iff max abs price diff &le; 0.05.</li>
 *   <li>Sensitivity — analytic BS delta vs central finite difference
 *       (h = 1e-4*S) on the same grid; PASS iff max diff &le; 1e-6.</li>
 *   <li>Stability — SBM capital recomputed with GIRR delta RWs x0.9 / x1.1;
 *       finding if |delta capital| / base capital &gt; 0.25.</li>
 *   <li>Backtesting / PLAT — desk zones from the IMA sketch.</li>
 *   <li>Data quality — staleness (&gt; 15 zero-change days) and gaps (any NaN).</li>
 * </ul>
 *
 * <p>Findings classification (pinned rule table): any High → verdict
 * "reject"; else any Medium → "approve-with-conditions"; else "approve".
 */
public final class Validation {

    private Validation() {
    }

    // ---- pinned check parameters -----------------------------------------
    /** Binomial benchmark lattice steps. */
    public static final int BENCH_STEPS = 501;
    /** Benchmark pass threshold on max |BS - binomial|. */
    public static final double BENCH_TOL = 0.05;
    /** Delta finite-difference pass threshold. */
    public static final double SENS_TOL = 1e-6;
    /** Capital stability threshold on |delta capital| / capital. */
    public static final double STABILITY_THRESHOLD = 0.25;
    /** Staleness threshold in zero-change days. */
    public static final int STALENESS_THRESHOLD = 15;

    private static final double[] BENCH_GRID_STRIKES = {70.0, 85.0, 100.0, 115.0, 130.0};
    private static final double[] BENCH_GRID_MATURITIES = {0.25, 0.5, 1.0, 2.0};
    private static final double BENCH_SPOT = 100.0;
    private static final double BENCH_RATE = 0.03;
    private static final double BENCH_DIV = 0.01;
    private static final double BENCH_VOL = 0.2;

    private static final Set<String> SEVERITIES = Set.of("High", "Medium", "Low");

    /** The ten pinned report section titles, in order. */
    public static final List<String> REPORT_SECTIONS = List.of(
            "1. Scope & Overview",
            "2. Pricing Benchmark",
            "3. Sensitivity Verification",
            "4. Capital Stability",
            "5. VaR Backtesting",
            "6. P&L Attribution (PLAT)",
            "7. Data Quality",
            "8. NMRF / SES",
            "9. Findings",
            "10. Overall Verdict");

    /**
     * One validation finding: pinned rule id, severity, human description.
     *
     * @param ruleId      pinned rule id, e.g. "BENCH-01"
     * @param severity    "High" | "Medium" | "Low"
     * @param description human-readable description
     */
    public record Finding(String ruleId, String severity, String description) {
        public Finding {
            if (!SEVERITIES.contains(severity)) {
                throw new IllegalArgumentException(
                        "Finding: severity must be one of " + SEVERITIES);
            }
        }
    }

    // ---------------------------------------------------------------------
    // Checks
    // ---------------------------------------------------------------------

    /** Max abs diff |BS - binomial(501)| over the pinned option grid. */
    public static double benchmarkMaxDiff() {
        double worst = 0.0;
        for (double k : BENCH_GRID_STRIKES) {
            for (double t : BENCH_GRID_MATURITIES) {
                for (boolean call : new boolean[] {true, false}) {
                    double a = Pricers.bsPrice(BENCH_SPOT, k, t, BENCH_RATE, BENCH_DIV,
                            BENCH_VOL, call);
                    double b = Pricers.binomialPrice(BENCH_SPOT, k, t, BENCH_RATE, BENCH_DIV,
                            BENCH_VOL, call, BENCH_STEPS);
                    worst = Math.max(worst, Math.abs(a - b));
                }
            }
        }
        return worst;
    }

    /** Max abs diff between analytic BS delta and a central finite difference. */
    public static double sensitivityMaxDiff() {
        double h = 1e-4 * BENCH_SPOT;
        double worst = 0.0;
        for (double k : BENCH_GRID_STRIKES) {
            for (double t : BENCH_GRID_MATURITIES) {
                for (boolean call : new boolean[] {true, false}) {
                    double analytic = Pricers.bsDelta(BENCH_SPOT, k, t, BENCH_RATE, BENCH_DIV,
                            BENCH_VOL, call);
                    double up = Pricers.bsPrice(BENCH_SPOT + h, k, t, BENCH_RATE, BENCH_DIV,
                            BENCH_VOL, call);
                    double dn = Pricers.bsPrice(BENCH_SPOT - h, k, t, BENCH_RATE, BENCH_DIV,
                            BENCH_VOL, call);
                    worst = Math.max(worst, Math.abs(analytic - (up - dn) / (2.0 * h)));
                }
            }
        }
        return worst;
    }

    /**
     * Data-quality stats of one series.
     *
     * @param staleDays number of zero-change days (NaNs excluded)
     * @param gaps      number of missing (NaN) values
     */
    public record DataQuality(int staleDays, int gaps) {
    }

    /** Staleness (# zero-change days) and gaps (# NaN values) of one series. */
    public static DataQuality dataQuality(double[] series) {
        if (series.length < 2) {
            throw new IllegalArgumentException("dataQuality: need at least 2 observations");
        }
        int gaps = 0;
        List<Double> clean = new ArrayList<>();
        for (double v : series) {
            if (Double.isNaN(v)) {
                gaps++;
            } else {
                clean.add(v);
            }
        }
        int stale = 0;
        for (int i = 1; i < clean.size(); i++) {
            if (clean.get(i) - clean.get(i - 1) == 0.0) {
                stale++;
            }
        }
        return new DataQuality(stale, gaps);
    }

    // ---------------------------------------------------------------------
    // Findings classification (pinned rule table)
    // ---------------------------------------------------------------------

    /**
     * Everything the pinned rule table needs to classify one desk.
     *
     * @param benchmarkMaxDiff   max |BS - binomial| over the pinned grid
     * @param sensitivityMaxDiff max |analytic - FD| delta over the grid
     * @param stabilityRelChange max(|dCap x1.1|, |dCap x0.9|) / base capital
     * @param backtestZone       the desk's VaR backtest zone
     * @param platZone           the desk's PLAT zone
     * @param staleDays          zero-change days in the desk P&amp;L
     * @param gaps               missing values in the desk P&amp;L
     */
    public record DeskCheckInputs(double benchmarkMaxDiff, double sensitivityMaxDiff,
                                  double stabilityRelChange, String backtestZone,
                                  String platZone, int staleDays, int gaps) {
    }

    /** Apply the pinned rule table; returns findings in table order. */
    public static List<Finding> classifyFindings(DeskCheckInputs c) {
        List<Finding> out = new ArrayList<>();
        if (c.benchmarkMaxDiff() > BENCH_TOL) {
            out.add(new Finding("BENCH-01", "High", "Pricing benchmark max diff "
                    + g6(c.benchmarkMaxDiff()) + " exceeds tolerance " + BENCH_TOL));
        }
        if (c.sensitivityMaxDiff() > SENS_TOL) {
            out.add(new Finding("SENS-01", "High", "Analytic vs FD delta max diff "
                    + g6(c.sensitivityMaxDiff()) + " exceeds " + SENS_TOL));
        }
        if ("red".equals(c.backtestZone())) {
            out.add(new Finding("BT-01", "High", "VaR backtest in RED zone"));
        }
        if ("amber".equals(c.backtestZone())) {
            out.add(new Finding("BT-02", "Medium", "VaR backtest in AMBER zone"));
        }
        if ("red".equals(c.platZone())) {
            out.add(new Finding("PLAT-01", "High", "PLAT in RED zone"));
        }
        if ("amber".equals(c.platZone())) {
            out.add(new Finding("PLAT-02", "Medium", "PLAT in AMBER zone"));
        }
        if (c.stabilityRelChange() > STABILITY_THRESHOLD) {
            out.add(new Finding("STAB-01", "Medium", String.format(Locale.US,
                    "Capital moves %.1f%% under +/-10%% GIRR RW (threshold %.0f%%)",
                    c.stabilityRelChange() * 100.0, STABILITY_THRESHOLD * 100.0)));
        }
        if (c.staleDays() > STALENESS_THRESHOLD) {
            out.add(new Finding("DQ-01", "Medium", c.staleDays()
                    + " zero-change days exceed staleness threshold " + STALENESS_THRESHOLD));
        }
        if (c.gaps() > 0) {
            out.add(new Finding("DQ-02", "Low",
                    c.gaps() + " missing values in the P&L series"));
        }
        return out;
    }

    /** Pinned verdict rule: any High → reject; any Medium → approve-with-conditions. */
    public static String overallVerdict(List<Finding> findings) {
        boolean high = false;
        boolean medium = false;
        for (Finding f : findings) {
            high |= "High".equals(f.severity());
            medium |= "Medium".equals(f.severity());
        }
        if (high) {
            return "reject";
        }
        if (medium) {
            return "approve-with-conditions";
        }
        return "approve";
    }

    // ---------------------------------------------------------------------
    // Report generation (structured results -> markdown)
    // ---------------------------------------------------------------------

    private static String fmt(double x) {
        return String.format(Locale.US, "%,.2f", x);
    }

    /** Python-style {@code %.6g} (trailing zeros stripped). */
    static String g6(double v) {
        String s = String.format(Locale.US, "%.6g", v);
        String exp = "";
        int e = s.indexOf('e');
        if (e >= 0) {
            exp = s.substring(e);
            s = s.substring(0, e);
        }
        if (s.contains(".")) {
            s = s.replaceAll("0+$", "").replaceAll("\\.$", "");
        }
        return s + exp;
    }

    /**
     * Render the validation report markdown from the engine's results.
     * Always emits every section in {@link #REPORT_SECTIONS} (tested by
     * string-contains checks).
     */
    public static String renderReport(Map<String, Engine.DeskIma> ima,
                                      Engine.ValidationResults val) {
        List<String> desks = new ArrayList<>(new TreeSet<>(ima.keySet()));
        StringBuilder sb = new StringBuilder();

        sb.append("# Independent Model Validation Report\n\n");
        sb.append("> Educational FRTB implementation — Basel-2019-flavored pinned parameter set.\n");
        sb.append("> NOT a compliant capital engine; for teaching and testing only.\n\n");
        sb.append("## ").append(REPORT_SECTIONS.get(0)).append("\n\n");
        sb.append("Desks in scope: ").append(String.join(", ", desks))
                .append(". Framework: SBM + DRC + RRAO (SA) and ")
                .append("ES/IMCC + PLAT + backtesting + SES (IMA sketch).\n\n");

        sb.append("## ").append(REPORT_SECTIONS.get(1)).append("\n\n");
        sb.append("| metric | value | threshold | result |\n");
        sb.append("|---|---|---|---|\n");
        double bmd = val.benchmarkMaxDiff();
        sb.append(String.format(Locale.US,
                "| max abs diff BS vs binomial(%d) | %.3e | %s | %s |\n",
                BENCH_STEPS, bmd, BENCH_TOL, bmd <= BENCH_TOL ? "PASS" : "FAIL"));
        sb.append("\n");

        sb.append("## ").append(REPORT_SECTIONS.get(2)).append("\n\n");
        double smd = val.sensitivityMaxDiff();
        sb.append(String.format(Locale.US,
                "Analytic BS delta vs central finite difference: max abs diff %.3e "
                        + "(threshold %.0e) — %s.\n\n",
                smd, SENS_TOL, smd <= SENS_TOL ? "PASS" : "FAIL"));

        sb.append("## ").append(REPORT_SECTIONS.get(3)).append("\n\n");
        sb.append("| scenario | SBM capital | change vs base |\n");
        sb.append("|---|---|---|\n");
        double baseCap = val.stabilityBaseCapital();
        double upCap = val.stabilityCapitalRwUp10();
        double dnCap = val.stabilityCapitalRwDn10();
        sb.append("| base | ").append(fmt(baseCap)).append(" | — |\n");
        sb.append("| GIRR delta RW x1.1 | ").append(fmt(upCap)).append(" | ")
                .append(fmt(upCap - baseCap)).append(" |\n");
        sb.append("| GIRR delta RW x0.9 | ").append(fmt(dnCap)).append(" | ")
                .append(fmt(dnCap - baseCap)).append(" |\n\n");

        sb.append("## ").append(REPORT_SECTIONS.get(4)).append("\n\n");
        sb.append("| desk | exceptions | zone | multiplier |\n");
        sb.append("|---|---|---|---|\n");
        for (String d : desks) {
            Ima.BacktestResult bt = ima.get(d).backtest();
            sb.append(String.format(Locale.US, "| %s | %d | %s | %.2f |\n",
                    d, bt.exceptions(), bt.zone(), bt.multiplier()));
        }
        sb.append("\n");

        sb.append("## ").append(REPORT_SECTIONS.get(5)).append("\n\n");
        sb.append("| desk | spearman | KS | zone | surcharge |\n");
        sb.append("|---|---|---|---|---|\n");
        for (String d : desks) {
            Plat.PlatResult pl = ima.get(d).plat();
            String sp = pl.spearman() == null ? "n/a"
                    : String.format(Locale.US, "%.4f", pl.spearman());
            String ks = pl.ks() == null ? "n/a"
                    : String.format(Locale.US, "%.4f", pl.ks());
            sb.append("| ").append(d).append(" | ").append(sp).append(" | ").append(ks)
                    .append(" | ").append(pl.zone()).append(" | ")
                    .append(fmt(ima.get(d).platSurcharge())).append(" |\n");
        }
        sb.append("\n");

        sb.append("## ").append(REPORT_SECTIONS.get(6)).append("\n\n");
        sb.append("| desk | zero-change days | gaps |\n");
        sb.append("|---|---|---|\n");
        for (String d : desks) {
            DataQuality dq = val.dataQuality().get(d);
            sb.append("| ").append(d).append(" | ").append(dq.staleDays()).append(" | ")
                    .append(dq.gaps()).append(" |\n");
        }
        sb.append("\n");

        sb.append("## ").append(REPORT_SECTIONS.get(7)).append("\n\n");
        sb.append("| desk | SES |\n");
        sb.append("|---|---|\n");
        for (String d : desks) {
            sb.append("| ").append(d).append(" | ").append(fmt(ima.get(d).ses())).append(" |\n");
        }
        sb.append("\n");

        sb.append("## ").append(REPORT_SECTIONS.get(8)).append("\n\n");
        boolean anyFinding = false;
        for (String d : desks) {
            for (Finding f : val.findings().get(d)) {
                sb.append("- **").append(f.severity()).append("** [").append(f.ruleId())
                        .append("] (").append(d).append("): ").append(f.description())
                        .append("\n");
                anyFinding = true;
            }
        }
        if (!anyFinding) {
            sb.append("- No findings.\n");
        }
        sb.append("\n");

        sb.append("## ").append(REPORT_SECTIONS.get(9)).append("\n\n");
        for (String d : desks) {
            sb.append("- ").append(d).append(": **").append(val.verdicts().get(d))
                    .append("**\n");
        }
        return sb.toString();
    }
}

File: rust/src/validation.rs
//! Independent model validation framework.
//!
//! Checks (all pinned):
//!
//! * Benchmarking — project BS pricer vs an independent CRR binomial
//!   lattice (501 steps) on a pinned strike/maturity/call-put grid;
//!   PASS iff max abs price diff <= 0.05.
//! * Sensitivity — analytic BS delta vs central finite difference
//!   (`h = 1e-4 * S`) on the same grid; PASS iff max diff <= 1e-6.
//! * Stability — SBM capital recomputed with GIRR delta RWs x0.9 / x1.1;
//!   finding if `|delta capital| / base capital > 0.25`.
//! * Backtesting — desk VaR backtest zone (from [`crate::ima`]).
//! * PLAT — desk PLAT zone (from [`crate::plat`]).
//! * Data quality — staleness: # of zero-change days > 15 -> finding;
//!   gaps: any missing (NaN) value -> finding.
//!
//! Findings classification (pinned rule table, see [`classify_findings`]):
//! High -> verdict `reject`; Medium -> `approve-with-conditions` (if no
//! High); else `approve`.

use std::collections::BTreeMap;

use crate::error::{invalid, Result};
use crate::ima::DeskIma;
use crate::pricers::{binomial_price, bs_delta, bs_price};

/// Pinned binomial benchmark steps.
pub const BENCH_STEPS: usize = 501;
/// Pinned benchmark pass tolerance on |BS - binomial|.
pub const BENCH_TOL: f64 = 0.05;
/// Pinned sensitivity pass tolerance on |analytic - FD| delta.
pub const SENS_TOL: f64 = 1e-6;
/// Pinned stability threshold on |delta capital| / capital.
pub const STABILITY_THRESHOLD: f64 = 0.25;
/// Pinned staleness threshold (zero-change days).
pub const STALENESS_THRESHOLD: usize = 15;
/// Pinned benchmark grid strikes.
pub const BENCH_GRID_STRIKES: [f64; 5] = [70.0, 85.0, 100.0, 115.0, 130.0];
/// Pinned benchmark grid maturities.
pub const BENCH_GRID_MATURITIES: [f64; 4] = [0.25, 0.5, 1.0, 2.0];
/// Pinned benchmark spot / rate / dividend yield / vol.
pub const BENCH_SPOT: f64 = 100.0;
/// Pinned benchmark risk-free rate.
pub const BENCH_RATE: f64 = 0.03;
/// Pinned benchmark dividend yield.
pub const BENCH_DIV: f64 = 0.01;
/// Pinned benchmark volatility.
pub const BENCH_VOL: f64 = 0.2;

/// Finding severities, highest first.
pub const SEVERITIES: [&str; 3] = ["High", "Medium", "Low"];
/// Possible overall verdicts.
pub const VERDICTS: [&str; 3] = ["approve", "approve-with-conditions", "reject"];

/// One validation finding: pinned rule id, severity, human description.
#[derive(Debug, Clone, PartialEq)]
pub struct Finding {
    /// Rule identifier (e.g. "BENCH-01").
    pub rule_id: String,
    /// "High", "Medium" or "Low".
    pub severity: String,
    /// Human-readable description.
    pub description: String,
}

impl Finding {
    fn new(rule_id: &str, severity: &str, description: String) -> Finding {
        Finding { rule_id: rule_id.to_string(), severity: severity.to_string(), description }
    }
}

// --------------------------------------------------------------------------
// Checks
// --------------------------------------------------------------------------

/// Max abs diff |BS - binomial(501)| over the pinned option grid.
pub fn benchmark_max_diff() -> Result<f64> {
    let mut worst = 0.0f64;
    for k in BENCH_GRID_STRIKES {
        for t in BENCH_GRID_MATURITIES {
            for call in [true, false] {
                let a = bs_price(BENCH_SPOT, k, t, BENCH_RATE, BENCH_DIV, BENCH_VOL, call)?;
                let b = binomial_price(
                    BENCH_SPOT, k, t, BENCH_RATE, BENCH_DIV, BENCH_VOL, call, BENCH_STEPS,
                )?;
                worst = worst.max((a - b).abs());
            }
        }
    }
    Ok(worst)
}

/// Max abs diff between analytic BS delta and a central finite difference
/// (`h = 1e-4 * S`) over the pinned grid.
pub fn sensitivity_max_diff() -> Result<f64> {
    let h = 1e-4 * BENCH_SPOT;
    let mut worst = 0.0f64;
    for k in BENCH_GRID_STRIKES {
        for t in BENCH_GRID_MATURITIES {
            for call in [true, false] {
                let analytic =
                    bs_delta(BENCH_SPOT, k, t, BENCH_RATE, BENCH_DIV, BENCH_VOL, call)?;
                let up = bs_price(BENCH_SPOT + h, k, t, BENCH_RATE, BENCH_DIV, BENCH_VOL, call)?;
                let dn = bs_price(BENCH_SPOT - h, k, t, BENCH_RATE, BENCH_DIV, BENCH_VOL, call)?;
                worst = worst.max((analytic - (up - dn) / (2.0 * h)).abs());
            }
        }
    }
    Ok(worst)
}

/// Data-quality metrics of one series.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct DataQuality {
    /// Number of zero-change days between consecutive non-missing values.
    pub stale_days: usize,
    /// Number of missing (NaN) values.
    pub gaps: usize,
}

/// Staleness (# zero-change days) and gaps (# NaN values) of one series.
pub fn data_quality(series: &[f64]) -> Result<DataQuality> {
    if series.len() < 2 {
        return invalid("data_quality: need at least 2 observations");
    }
    let gaps = series.iter().filter(|v| v.is_nan()).count();
    let clean: Vec<f64> = series.iter().copied().filter(|v| !v.is_nan()).collect();
    let stale = clean.windows(2).filter(|w| w[1] - w[0] == 0.0).count();
    Ok(DataQuality { stale_days: stale, gaps })
}

// --------------------------------------------------------------------------
// Findings classification (pinned rule table)
// --------------------------------------------------------------------------

/// Everything the rule table needs to classify one desk.
#[derive(Debug, Clone, PartialEq)]
pub struct DeskCheckInputs {
    /// Benchmark check result.
    pub benchmark_max_diff: f64,
    /// Sensitivity (delta vs FD) check result.
    pub sensitivity_max_diff: f64,
    /// `max(|dCap(x1.1)|, |dCap(x0.9)|) / base capital`.
    pub stability_rel_change: f64,
    /// Desk backtest zone.
    pub backtest_zone: String,
    /// Desk PLAT zone.
    pub plat_zone: String,
    /// Zero-change days in the desk P&L.
    pub stale_days: usize,
    /// Missing values in the desk P&L.
    pub gaps: usize,
}

/// Apply the pinned rule table; returns findings in table order.
///
/// | rule | severity | fires when |
/// |---|---|---|
/// | BENCH-01 | High   | benchmark max diff > 0.05 |
/// | SENS-01  | High   | delta FD max diff > 1e-6 |
/// | BT-01    | High   | backtest zone red |
/// | BT-02    | Medium | backtest zone amber |
/// | PLAT-01  | High   | PLAT zone red |
/// | PLAT-02  | Medium | PLAT zone amber |
/// | STAB-01  | Medium | max abs rel capital move > 0.25 |
/// | DQ-01    | Medium | zero-change days > 15 |
/// | DQ-02    | Low    | any missing value |
///
/// All comparisons are strict `>`.
pub fn classify_findings(c: &DeskCheckInputs) -> Vec<Finding> {
    let mut out = Vec::new();
    if c.benchmark_max_diff > BENCH_TOL {
        out.push(Finding::new(
            "BENCH-01",
            "High",
            format!(
                "Pricing benchmark max diff {} exceeds tolerance {BENCH_TOL}",
                fmt_g6(c.benchmark_max_diff)
            ),
        ));
    }
    if c.sensitivity_max_diff > SENS_TOL {
        out.push(Finding::new(
            "SENS-01",
            "High",
            format!(
                "Analytic vs FD delta max diff {} exceeds {SENS_TOL}",
                fmt_g6(c.sensitivity_max_diff)
            ),
        ));
    }
    if c.backtest_zone == "red" {
        out.push(Finding::new("BT-01", "High", "VaR backtest in RED zone".to_string()));
    }
    if c.backtest_zone == "amber" {
        out.push(Finding::new("BT-02", "Medium", "VaR backtest in AMBER zone".to_string()));
    }
    if c.plat_zone == "red" {
        out.push(Finding::new("PLAT-01", "High", "PLAT in RED zone".to_string()));
    }
    if c.plat_zone == "amber" {
        out.push(Finding::new("PLAT-02", "Medium", "PLAT in AMBER zone".to_string()));
    }
    if c.stability_rel_change > STABILITY_THRESHOLD {
        out.push(Finding::new(
            "STAB-01",
            "Medium",
            format!(
                "Capital moves {:.1}% under +/-10% GIRR RW (threshold {:.0}%)",
                c.stability_rel_change * 100.0,
                STABILITY_THRESHOLD * 100.0
            ),
        ));
    }
    if c.stale_days > STALENESS_THRESHOLD {
        out.push(Finding::new(
            "DQ-01",
            "Medium",
            format!(
                "{} zero-change days exceed staleness threshold {STALENESS_THRESHOLD}",
                c.stale_days
            ),
        ));
    }
    if c.gaps > 0 {
        out.push(Finding::new(
            "DQ-02",
            "Low",
            format!("{} missing values in the P&L series", c.gaps),
        ));
    }
    out
}

/// Pinned verdict rule: any High -> `reject`; any Medium ->
/// `approve-with-conditions`; else `approve`.
pub fn overall_verdict(findings: &[Finding]) -> &'static str {
    if findings.iter().any(|f| f.severity == "High") {
        "reject"
    } else if findings.iter().any(|f| f.severity == "Medium") {
        "approve-with-conditions"
    } else {
        "approve"
    }
}

// --------------------------------------------------------------------------
// Validation summary + report generation
// --------------------------------------------------------------------------

/// All validation outputs for the report and the golden suite.
#[derive(Debug, Clone, PartialEq)]
pub struct ValidationSummary {
    /// Benchmark check result.
    pub benchmark_max_diff: f64,
    /// Delta-vs-FD check result.
    pub sensitivity_max_diff: f64,
    /// Firm SBM capital under the base parameter set.
    pub stability_base_capital: f64,
    /// Firm SBM capital under GIRR delta RW x1.1.
    pub stability_capital_rw_up10: f64,
    /// Firm SBM capital under GIRR delta RW x0.9.
    pub stability_capital_rw_dn10: f64,
    /// `max(|dCap up|, |dCap dn|) / base` (0 when base is 0).
    pub stability_rel_change: f64,
    /// Per-desk data-quality metrics.
    pub data_quality: BTreeMap<String, DataQuality>,
    /// Per-desk findings, in rule-table order.
    pub findings: BTreeMap<String, Vec<Finding>>,
    /// Per-desk overall verdicts (exact pinned strings).
    pub verdicts: BTreeMap<String, String>,
    /// The rendered markdown report.
    pub report_md: String,
}

/// The ten pinned report section titles (always all emitted, prefixed
/// `## `).
pub const REPORT_SECTIONS: [&str; 10] = [
    "1. Scope & Overview",
    "2. Pricing Benchmark",
    "3. Sensitivity Verification",
    "4. Capital Stability",
    "5. VaR Backtesting",
    "6. P&L Attribution (PLAT)",
    "7. Data Quality",
    "8. NMRF / SES",
    "9. Findings",
    "10. Overall Verdict",
];

/// Format with thousands separators and a fixed number of decimals
/// (Python's `{:,.2f}` style).
pub fn fmt_thousands(x: f64, decimals: usize) -> String {
    let s = format!("{:.*}", decimals, x);
    let (sign, rest) = match s.strip_prefix('-') {
        Some(r) => ("-", r),
        None => ("", s.as_str()),
    };
    let (int_part, frac_part) = match rest.split_once('.') {
        Some((i, f)) => (i, Some(f)),
        None => (rest, None),
    };
    let mut grouped = String::new();
    let bytes = int_part.as_bytes();
    for (i, b) in bytes.iter().enumerate() {
        if i > 0 && (bytes.len() - i) % 3 == 0 {
            grouped.push(',');
        }
        grouped.push(*b as char);
    }
    match frac_part {
        Some(f) => format!("{sign}{grouped}.{f}"),
        None => format!("{sign}{grouped}"),
    }
}

/// Scientific notation with 3 decimals and a signed two-digit exponent
/// (Python's `{:.3e}` style, e.g. `5.163e-03`).
pub fn fmt_sci3(x: f64) -> String {
    let s = format!("{:.3e}", x);
    match s.split_once('e') {
        Some((mant, exp)) => {
            let e: i32 = exp.parse().unwrap_or(0);
            format!("{mant}e{}{:02}", if e < 0 { "-" } else { "+" }, e.abs())
        }
        None => s,
    }
}

/// Python `%g`-flavored formatting with 6 significant digits (used in
/// finding descriptions).
pub fn fmt_g6(x: f64) -> String {
    if x == 0.0 {
        return "0".to_string();
    }
    let exp = x.abs().log10().floor() as i32;
    if (-4..6).contains(&exp) {
        let decimals = (5 - exp).max(0) as usize;
        let mut s = format!("{:.*}", decimals, x);
        if s.contains('.') {
            s = s.trim_end_matches('0').trim_end_matches('.').to_string();
        }
        s
    } else {
        let mut mant = format!("{:.5e}", x);
        if let Some((m, e)) = mant.clone().split_once('e') {
            let m = m.trim_end_matches('0').trim_end_matches('.');
            let ev: i32 = e.parse().unwrap_or(0);
            mant = format!("{m}e{}{:02}", if ev < 0 { "-" } else { "+" }, ev.abs());
        }
        mant
    }
}

fn fmt(x: f64) -> String {
    fmt_thousands(x, 2)
}

/// Render the validation report markdown from the engine's results.
///
/// Always emits every section in [`REPORT_SECTIONS`] (tested by
/// string-contains checks in every language).
pub fn render_report(ima: &BTreeMap<String, DeskIma>, val: &ValidationSummary) -> String {
    let desks: Vec<&String> = ima.keys().collect();
    let mut lines: Vec<String> = Vec::new();
    let mut add = |s: String| lines.push(s);

    add("# Independent Model Validation Report".to_string());
    add(String::new());
    add("> Educational FRTB implementation — Basel-2019-flavored pinned parameter set.".to_string());
    add("> NOT a compliant capital engine; for teaching and testing only.".to_string());
    add(String::new());
    add(format!("## {}", REPORT_SECTIONS[0]));
    add(String::new());
    let desk_list: Vec<String> = desks.iter().map(|d| d.to_string()).collect();
    add(format!(
        "Desks in scope: {}. Framework: SBM + DRC + RRAO (SA) and ES/IMCC + PLAT + backtesting + SES (IMA sketch).",
        desk_list.join(", ")
    ));
    add(String::new());

    add(format!("## {}", REPORT_SECTIONS[1]));
    add(String::new());
    add("| metric | value | threshold | result |".to_string());
    add("|---|---|---|---|".to_string());
    let bmd = val.benchmark_max_diff;
    add(format!(
        "| max abs diff BS vs binomial({BENCH_STEPS}) | {} | {BENCH_TOL} | {} |",
        fmt_sci3(bmd),
        if bmd <= BENCH_TOL { "PASS" } else { "FAIL" }
    ));
    add(String::new());

    add(format!("## {}", REPORT_SECTIONS[2]));
    add(String::new());
    let smd = val.sensitivity_max_diff;
    // Python renders the 1e-6 threshold as "1e-06" (two-digit exponent).
    let sens_tol = {
        let s = format!("{SENS_TOL:e}");
        match s.split_once('e') {
            Some((m, e)) => {
                let ev: i32 = e.parse().unwrap_or(0);
                format!("{m}e{}{:02}", if ev < 0 { "-" } else { "+" }, ev.abs())
            }
            None => s,
        }
    };
    add(format!(
        "Analytic BS delta vs central finite difference: max abs diff {} (threshold {sens_tol}) — {}.",
        fmt_sci3(smd),
        if smd <= SENS_TOL { "PASS" } else { "FAIL" }
    ));
    add(String::new());

    add(format!("## {}", REPORT_SECTIONS[3]));
    add(String::new());
    add("| scenario | SBM capital | change vs base |".to_string());
    add("|---|---|---|".to_string());
    let (base_cap, up_cap, dn_cap) = (
        val.stability_base_capital,
        val.stability_capital_rw_up10,
        val.stability_capital_rw_dn10,
    );
    add(format!("| base | {} | — |", fmt(base_cap)));
    add(format!("| GIRR delta RW x1.1 | {} | {} |", fmt(up_cap), fmt(up_cap - base_cap)));
    add(format!("| GIRR delta RW x0.9 | {} | {} |", fmt(dn_cap), fmt(dn_cap - base_cap)));
    add(String::new());

    add(format!("## {}", REPORT_SECTIONS[4]));
    add(String::new());
    add("| desk | exceptions | zone | multiplier |".to_string());
    add("|---|---|---|---|".to_string());
    for d in &desks {
        let bt = &ima[*d].backtest;
        add(format!("| {d} | {} | {} | {:.2} |", bt.exceptions, bt.zone, bt.multiplier));
    }
    add(String::new());

    add(format!("## {}", REPORT_SECTIONS[5]));
    add(String::new());
    add("| desk | spearman | KS | zone | surcharge |".to_string());
    add("|---|---|---|---|---|".to_string());
    for d in &desks {
        let pl = &ima[*d].plat;
        let sp = pl.spearman.map_or("n/a".to_string(), |v| format!("{v:.4}"));
        let ks = pl.ks.map_or("n/a".to_string(), |v| format!("{v:.4}"));
        add(format!("| {d} | {sp} | {ks} | {} | {} |", pl.zone, fmt(ima[*d].plat_surcharge)));
    }
    add(String::new());

    add(format!("## {}", REPORT_SECTIONS[6]));
    add(String::new());
    add("| desk | zero-change days | gaps |".to_string());
    add("|---|---|---|".to_string());
    for d in &desks {
        let dq = &val.data_quality[*d];
        add(format!("| {d} | {} | {} |", dq.stale_days, dq.gaps));
    }
    add(String::new());

    add(format!("## {}", REPORT_SECTIONS[7]));
    add(String::new());
    add("| desk | SES |".to_string());
    add("|---|---|".to_string());
    for d in &desks {
        add(format!("| {d} | {} |", fmt(ima[*d].ses)));
    }
    add(String::new());

    add(format!("## {}", REPORT_SECTIONS[8]));
    add(String::new());
    let mut any_finding = false;
    for d in &desks {
        for f in &val.findings[*d] {
            add(format!("- **{}** [{}] ({d}): {}", f.severity, f.rule_id, f.description));
            any_finding = true;
        }
    }
    if !any_finding {
        add("- No findings.".to_string());
    }
    add(String::new());

    add(format!("## {}", REPORT_SECTIONS[9]));
    add(String::new());
    for d in &desks {
        add(format!("- {d}: **{}**", val.verdicts[*d]));
    }
    add(String::new());
    lines.join("\n")
}
